### DISCUSSION TITLE
Key rotation

### DISCUSSION THREAD

**Discussion by overheadhunter** [encryption-alliance/unified-vault-format#2] [Approved]
I would like to propose an idea allowing to revoke access to an encrypted file structure.

There are situations, e.g. after changing a password or revoking access to a vault, when we want to cryptographically ensure, that future changes to vault contents may not be accessible to anyone who has been in possession of the key in the past. This leads to requiring a key rotation mechanism:

In our vault metadata file, we need to keep track of key revisions. Any new file shall be encrypted with the latest revision. Existing files don't need to be re-encrypted as long as they are not changed, since we can assume that anyone owning the old key could have copied the decrypted file anyway.

It is obviously impossible to destroy data, once it has been shared. But we can reliably protect the confidentiality of any future changes ("perfect forward secrecy" style).

**Comment by overheadhunter** [comment]
It should be noted, that anyone in possession of an old key is technically still able to add files using his old key. Well-behaving applications may not allow this, but there is not cryptographic protection against tampering files without notice. An attack may look like this:

1. Alice stores "be nice!" in `instructions.txt` using the latest key revision 13.
2. Mallory (who was invited to collaborate until revision 5) will now overwrite `instructions.txt` with "destroy all humans!" using key revision 5.
3. Bob the bot will now read `instructions.txt` and execute them, since the contents are technically authentic

**Reply by overheadhunter** [reply]
if we store the date of the key rotation event, software could check the last-modified date of the encrypted file and issue a warning if encountering files that have been modified with a then outdated key.

**Comment by overheadhunter** [comment]
Race conditions:

What happens in case of concurrent key rotations? Let's assume that for some existing vault that Alice and Bob have access to:

1. in parallel on two different machines
    * Alice changes the passphrase and then adds `a.txt`
    * Bob changes the passphrase and then adds `b.txt`
2. sync! 💥 
3. assume both Alice and Bob resolve the conflicting vault config file with the incoming change
4. consequently neither knows the new key rotated by the other party, i.e. Alice can not decrypt `b.txt` and Bob can not decrypt `a.txt` (which is correct behavior)
5. due to step 3. both also lost access to their "own" file

How can we resolve this?

**Reply by overheadhunter** [reply]
One option might be to instead of saving the key revision history in the vault config file, store it in multiple files inside some `keys/` subdirectory.

Still holds the potential for diverging timelines, but syncs without conflict, so software can handle merging.

**Reply by chenkins** [reply]
@overheadhunter upon key rotation, would the files inside the `keys/` subdirectory would all have to be re-encrypted?

**Reply by overheadhunter** [reply]
> @overheadhunter upon key rotation, would the files inside the `keys/` subdirectory would all have to be re-encrypted?

No, a new file would be added, the existing ones should remain untouched.

Example:

1. Initially, `keys/` is empty. The latest key only exists in-memory.
2. Upon key rotation, `keys/key1` is added, which is the key from step 1, encrypted using the new key (which only exists in-memory)
3. repeat.

So this dir only holds a "linked list" of outdated keys, the in-use key is not yet part of the collection.

**Reply by overheadhunter** [reply]
We can **not avoid conflicts**: Even when using individual files, the metadata file located in the root directory must always be encrypted using the _latest_ key, thus the same file is being updated during concurrent key rotations.
